# 221HW1
**Bohao Zou**  917796070  
**Bingdao Chen**  917781027  
**date**: 05/12/2020

__Pledge:__  
Please sign below (print full name) after checking the following. If you can not honestly check each of these responses, please email me at kbala@ucdavis.edu to explain your situation.  

We pledge that we are honest students with academic integrity and we have not cheated on this homework.

These answers are our own work.

We did not give any other students assistance on this homework.

We understand that to submit work that is not our own and pretend that it is our is a violation of the UC Davis code of conduct and will be reported to Student Judicial Affairs.

We understand that suspected misconduct on this homework will be reported to the Office of Student Support and Judicial Affairs and, if established, will result in disciplinary sanctions up through Dismissal from the University and a grade penalty up to a grade of ``F" for the course.



Team Member 1:  Bingdao Chen  
Team Member 2:  Bohao Zou


## Question 1
Answer true or false for each of the question below and give justification. 

(a) Gradient descent is a supervised learning algorithm. 

(b) Gradient descent is an algorithm to minimize or maximize a function.  

(c) Logistic regression cannot be performed after linear PCA.  

(d) Support vector machine is non-linear classification algorithm.  

(e) A regression algorithm can be modified to be used for classification as well.  

### Solution
(a) False. Supervised Learning consists of two problems: Regression and Classfication. Gradient descent is an optimization algorithm used to find the minimizer of functions.

(b) True. Gradient descent is an algorithm to minimize a function. For the maximization problem of function $f(x)$, we can transfer it to find the minimizer of the $-f(x)$.

(c) False. PCA is to reduce the dimentionality of the data by coming up with linear combinations of the variables that maximize the variance explained. Logistic regression can also be performed on thses newly masde variables.

(d) False. Support vector machine is linear classification algorithm, because it creates a line or a hyperplane which separates the data into classes. However, SVM with kernel trick is a non-linear classification for mapping the data into the high-dimensional space.

(e) True. We can add some functions to the output of regression like sigmoid or softmax to transform the regression problem to classification problem

## Qustion 2
### Solution
#### (a)

The **Kullback-Leibler (KL) divergence** is defined as 
$$
D_{KL}(P\|Q) = \mathbb{E}_{x\sim P}\left[log\frac{P(x)}{Q(x)}\right] = \mathbb{E}_{x\sim P}\left[logP(x)-logQ(x)\right]
$$
It is used to measure how different two distributions if there are two separate probability distributions $P(x)$ and $Q(x)$ over the same random variablex. Therein, the KL divergence is non-negative and asymmetry, so the sequence of distributions is important. For the issue that using a distribution $q(x)$ to approximate the distribution $p(x)$, the results are different if minimizing $D_{KL}(p\|q)$ and $D_{KL}(q\|p)$. For the first situation, $q(x)$ are more likely to  put high probability mass on all of modes of $p(x)$. For the second situation, $q(x)$ is more focused on one single mode of $p(x)$, avoiding putting probability mass in the low probability areas between modes of $p(x)$.

#### (b)

For an unkonwn data-generating distribution $p_{data}(x)$, we want to find a good estimator $p_{model}(x;\theta)$ for the true probability $p_{data}(x)$. Maximum Likelihood Estimation (MLE) is one method, and it is defined as 
$$
\theta_{MLE} = \mathrm{argmax}_{\theta}\prod_{i=1}^m p_{model}(x^{(i)};\theta)
$$
After log-transformation and rescaling, it can be expressed as an expectation with respect to the empirical distribution $\hat p_{data}$:
$$
\theta_{MLE} = \mathrm{argmax}_{\theta}\mathbb{E}_{x\sim\hat p_{data}}\mathrm{log}p_{model}(x;\theta)
$$
Besides finding the maximum likelihood $\theta_{MLE}$, another way to view this problem is to minimize the dissimilarity between the empirical distribution $\hat p_{data}$ and the model distribution $p_{model}$, which can be measured by KL divergence as follows.
$$
D_{KL}(\hat p_{data}\| p_{model}) = \mathbb{E}_{x\sim\hat p_{data}}\left[\mathrm{log}\hat p_{data}(x) - \mathrm{log}p_{model}(x) \right]
$$
Considering $\mathrm{log}\hat p_{data}(x)$ is finxed, so this problem is equivalent to minimizing $-\mathbb{E}_{x\sim\hat p_{data}}[\mathrm{log}p_{model}(x)]$, which is same as the above MLE equation.

In order to obtain a good estimator, we can either maximize the likelihood or minimize KL divergence. However, in practice, minimizing KL divergence is more desired for its non-negative property.

#### (c)

Some issues with explicit density models:

(1) Long running time: For instance, FVBNs cannot generate samples in parallel, whose runtime is proportional to the dimensionality of $x$.

(2) Too many restrictions: For instance, only a few probability distributions adimt tractable Markov chain sampling in Boltzmann machines. Generator must be invertible and the latent code $z$ must have the same dimension as the samples $x$ in non-linear ICA.

(3) The need of Markov chains: For instance, Boltzmann machines and GSNs.

(4) The need of variational bound and asymptotically consistent: For instance, some VAEs, but it is not yet proven.

(5) The quality of samples produced by explicit density models is worse than GANs.

#### (d)

The main idea behind GANs:

The framework of GANs can be interpreted as a game between two players. One is called discriminator and one is called generator. The discriminator examines samples from generator whether they are real or fake, and the generator is trained to fool the discriminator. In this process, the generator can learn to creat samples that are drawn from the same distribution as the training data.

Formally, GANs contains latent variables $z$ and observed variables $x$. Function $D$ and $G$ stands for the discriminator and generator and use $\theta^{(D)}$ and $\theta^{(G)}$ as parameters respectively. The purpose of both two is to minimize its own cost fucntion $J^{(D)}(\theta^{(D)},\theta^{(G)})$ and $J^{(G)}(\theta^{(D)},\theta^{(G)})$ respectively. Both of them can only control its own parameters, and during this process, the solution is called Nash equilibrium which is $(\theta^{(D)},\theta^{(G)})$. It is a local minimum of $J^{(D)}$ with respect to $\theta^{(D)}$ and a local minimum $J^{(G)}$ with respect to $\theta^{(G)}$.

#### (e)

Three drawbacks for GANs:

(1) It takes a lot time to train and needs huge computation.

(2) Training a GAN need to find a Nash equilibrium of a game. However, sometimes gradient descent can do this, sometimes it can not. We do not have a good algorithem for finding equilibrium.

(3) It’s hard to learn to generate discrete data, like text.

### Question 3
#### (a)

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

In [2]:
# read train and test data
house_train_data = pd.read_table('./housing_train.txt',sep= '\s+', header = None)
house_train_data.columns = ["x1", "x2", "x3", "x4", "x5","x6","x7","x8","x9","x10","x11","x12","x13","y"]
house_test_data = pd.read_table('./housing_test.txt',sep= '\s+', header = None)
house_test_data.columns = ["x1", "x2", "x3", "x4", "x5","x6","x7","x8","x9","x10","x11","x12","x13","y"]

# data pre-processing
y_tr = house_train_data['y'].values
X_tr = house_train_data.drop(['y'],axis=1).values
y_te = house_test_data['y'].values
X_te = house_test_data.drop(['y'],axis=1).values

In [3]:
# fit the model on train data and predict 
lrg = LinearRegression(fit_intercept = False).fit(X_tr, y_tr)

# mean square prediction error on train data 
mspe_tr = np.mean((y_tr - lrg.predict(X_tr))**2)
print("The mean square prediction error based on train data is " + str(mspe_tr) + ".")

The mean square prediction error based on train data is 24.475882784643677.


#### (b)

In [151]:
# mean square prediction error
mspe_te = np.mean((y_te - lrg.predict(X_te))**2)
print("The mean square prediction error based on test data is " + str(mspe_te) + ".")

The mean square prediction error based on test data is 24.292238175660525.


#### (c)

In [153]:
# obtain all X features 
X_poly_tr = PolynomialFeatures(degree = 2, include_bias = False).fit_transform(X_tr)
X_poly_te = PolynomialFeatures(degree = 2, include_bias = False).fit_transform(X_te)

# fit the model on the new train data 
lrg_poly = LinearRegression(fit_intercept = False).fit(X_poly_tr, y_tr)

# mean square prediction error on the new train data 
mspe_poly_tr = np.mean((y_tr - lrg_poly.predict(X_poly_tr))**2)
print("The mean square prediction error based on the new train data is " + str(mspe_poly_tr) + ".")

# mean square prediction error on the new test data 
mspe_poly_te = np.mean((y_te - lrg_poly.predict(X_poly_te))**2)
print("The mean square prediction error based on the new test data is " + str(mspe_poly_te) + ".")

The mean square prediction error based on the new train data is 5.431632129079359.
The mean square prediction error based on the new test data is 34.68642969650256.


Compared with the previous model, the mean squared prediction error of the polynomial linear regression model on train data is smaller, which indicates that it fits better. However,  the mean squared prediction error of the polynomial linear regression model on test data gets greater, which indicates that the polynomial model results in overfitting.  

### Question 4
#### (a)

In [22]:
import pyreadr
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.feature_extraction.text import CountVectorizer
from nltk.stem.snowball import FrenchStemmer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.dummy import DummyClassifier
import statistics

In [3]:
# load the data 
Amazon = pyreadr.read_r('Amazon.RData') 
df = Amazon["dat"]

In [4]:
np.unique(df["rating"])
print("For rating value 1, there are " +  str(sum(df["rating"]==1)) + " reviewers.")
print("For rating value 5, there are " +  str(sum(df["rating"]==5)) + " reviewers.")

For rating value 1, there are 656 reviewers.
For rating value 5, there are 656 reviewers.


The best performance of a constant classifier is that assigning all the reviews to one class which has the highest probability in the data. In this case, because the number of rating $= 5$ and rating $= 1$ are same, the misclassfication rate is 0.5.

#### (b)

In [5]:
# creat the Document-Term matrix and TF-IDF matrix
text = df.loc[:,"review"]

stemmer =FrenchStemmer()
analyzer= CountVectorizer().build_analyzer()
def stemmed_words(doc): 
    return (stemmer.stem(w) for w in analyzer(doc))


In [6]:
# Doucument term matrix generation
vec = CountVectorizer(token_pattern = '[a-z]{3,15}', analyzer = stemmed_words)
Xdtm = vec.fit_transform(text)
dtm_name = vec.get_feature_names()
dtm = pd.DataFrame(Xdtm.toarray(), columns = vec.get_feature_names())
print(dtm.shape)

# Tf-idf matrix generation
vectorizer = TfidfVectorizer(token_pattern = '[a-z]{3,15}', analyzer = stemmed_words)
Xtf = vectorizer.fit_transform(text)
tf_name = vectorizer.get_feature_names()
Tfidf = pd.DataFrame(Xtf.toarray(), columns = vectorizer.get_feature_names())
print(Tfidf.shape)

(1312, 5186)
(1312, 5186)


In [7]:
# standardize the data 
Xdtm_tr, Xdtm_te, ydtm_tr, ydtm_te = train_test_split(dtm, df['rating'], test_size=0.3, random_state=99)
Xtf_tr, Xtf_te, ytf_tr, ytf_te = train_test_split(Tfidf, df['rating'], test_size=0.3, random_state=99)

Because the regularization penalty is comprised of the sum of the absolute value of the coefficients, we need to scale the data such that its coefficients are all based on the same scale.

In [8]:
# standardize the data 
sc_dtm = StandardScaler()
Xdtm_tr_sd = sc_dtm.fit_transform(Xdtm_tr)
Xdtm_te_sd = sc_dtm.transform(Xdtm_te)

sc_tf = StandardScaler()
Xtf_tr_sd = sc_tf.fit_transform(Xtf_tr)
Xtf_te_sd = sc_tf.transform(Xtf_te)

In [9]:
# Fit a L1-regularized logistic regression model
dtm_lr = LogisticRegression(penalty='l1', solver='liblinear')
dtm_lr.fit(Xdtm_tr_sd, ydtm_tr)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l1',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

In [10]:
tf_lr = LogisticRegression(penalty='l1', solver='liblinear')
tf_lr.fit(Xtf_tr_sd, ytf_tr)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
                   intercept_scaling=1, l1_ratio=None, max_iter=100,
                   multi_class='auto', n_jobs=None, penalty='l1',
                   random_state=None, solver='liblinear', tol=0.0001, verbose=0,
                   warm_start=False)

#### (c)

In [11]:
print("There are " + str(sum(dtm_lr.coef_[0] != 0)) + 
      " covariates have non-zero coefficients in Document-Term Matrix.")
print("There are " + str(sum(tf_lr.coef_[0] != 0)) + 
      " covariates have non-zero coefficients in Tf-idf Matrix.")

There are 426 covariates have non-zero coefficients in Document-Term Matrix.
There are 457 covariates have non-zero coefficients in Tf-idf Matrix.


In [12]:
# verify that the most largest 20 coefficients are positive and the most least 20 coefficients are negative
print(dtm_lr.coef_[0][dtm_lr.coef_[0].argsort()[::-1][:20]] > 0)
print(dtm_lr.coef_[0][dtm_lr.coef_[0].argsort()[:20]] < 0)
print(tf_lr.coef_[0][tf_lr.coef_[0].argsort()[::-1][:20]] > 0)
print(tf_lr.coef_[0][tf_lr.coef_[0].argsort()[:20]] < 0)

[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True]
[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True]
[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True]
[ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True  True  True  True]


For Document-Term Matrix, the twenty words with the most positive coefficients are 

In [13]:
np.array(dtm_name)[dtm_lr.coef_[0].argsort()[::-1][:20]]

array(['lov', 'teething', 'soph', 'loved', 'voic', 'easy', 'great', 'she',
       'highly', 'teeth', 'gift', 'chew', 'hear', 'seen', 'sensit',
       'sony', 'channel', 'excellent', 'littl', 'awesom'], dtype='<U20')

For Document-Term Matrix, the twenty words with the most negative coefficients are twenty words are

In [14]:
np.array(dtm_name)[dtm_lr.coef_[0].argsort()[:20]]

array(['not', 'leak', 'returned', 'something', 'out', 'wast', 'off',
       'stopped', 'after', 'useless', 'get', 'back', 'return', 'even',
       'bin', 'becaus', 'doesn', 'disappointed', 'leaked', 'pictur'],
      dtype='<U20')

For Tf-idf Matrix, the twenty words with the most positive coefficients are

In [15]:
np.array(tf_name)[tf_lr.coef_[0].argsort()[::-1][:20]]

array(['lov', 'teething', 'soph', 'loved', 'she', 'voic', 'great',
       'highly', 'gift', 'toy', 'easy', 'sensit', 'littl', 'channel',
       'teeth', 'other', 'her', 'confined', 'hear', 'sony'], dtype='<U20')

For Tf-idf Matrix, the twenty words with the most negative coefficients are

In [16]:
np.array(tf_name)[tf_lr.coef_[0].argsort()[:20]]

array(['not', 'wast', 'off', 'out', 'diap', 'after', 'leak', 'bottl',
       'the', 'swing', 'wrap', 'horribl', 'motorol', 'return', 'beeping',
       'something', 'doesn', 'you', 'returned', 'pictur'], dtype='<U20')

#### (d)

In [17]:
print("The misclassification rate of Document-Term model is " + str(statistics.mean(ydtm_te != dtm_lr.predict(Xdtm_te_sd)))+ ".")
print("The misclassification rate of Tf-idf model is " + str(statistics.mean(ytf_te != tf_lr.predict(Xtf_te_sd)))+ ".")

The misclassification rate of Document-Term model is 0.07360406091370558.
The misclassification rate of Tf-idf model is 0.06598984771573604.


In [25]:
# misclassification rate for constant classifier
dummy_dtm = DummyClassifier(strategy="most_frequent")
dummy_dtm.fit(Xdtm_te, ydtm_te)
print(dummy_dtm.predict(Xdtm_te)[0])
print(1-dummy_dtm.score(Xdtm_te, ydtm_te))

dummy_tf = DummyClassifier(strategy="most_frequent")
dummy_tf.fit(Xtf_te, ytf_te)
print(dummy_tf.predict(Xtf_te)[0])
print(1-dummy_tf.score(Xtf_te, ytf_te))


1
0.47969543147208127
1
0.47969543147208127


From the result above, the misclassification rate of both model are very low. However, for the constant classifier, because it ignores the content of reviews and blindly assigns all reviews to one class, the results of document term matrix and Tf-idf matrix are same. It assigns all the review to the "bad" class, and the misclassfication rate is 0.4797, which is much higher than the misclassfication rate of the first model. The logisitic model is better.